Remember editor window mode, screen, size and position on restart

spanzeri opened the pull request:
Second attempt at a change to allow godot editor to remember the last size, screen, position and window mode across restarts.
When working on a godot project on a high res screen or with an external text editor, people might prefer to not have the editor maximized on screen.
The previous behaviour was for the editor window to start maximized, and when windowed it would resize to a very small size.
I made a previous attempt here: [https://github.com/godotengine/godot/pull/76076](https://github.com/godotengine/godot/pull/76076)
However, this version has some nice improvements:

- It runs earlier in the process, which means the window starts at the right size and does not change after the editor has loaded;
- Is per-project, so people can setup different work environments per project (which is also the behaviour for layouts)

The project setting defaults to false and is hidden under the advanced tab. This can be changed depending of what people decide the best "default behaviour" would be
<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->
*Bugsquad edit:* Closes https://github.com/godotengine/godot-proposals/issues/8358

spanzeri committed:
Remember editor window mode, screen, size and position on restart

Calinou commented on main/main.cpp:
```diff
@@ -1977,6 +1977,37 @@ Error Main::setup2(Thread::ID p_main_tid_override) {
 	}
+
+	if (editor) {
+		bool restore_editor_window = GLOBAL_DEF("editor/run/remember_window_size_and_position", false);
```
I think this behavior should be enabled by default. Most people expect this to happen by default, judging by the issues and proposals that were opened over the years.

spanzeri replied:
I agree, but I decided to err on the side of caution by leaving the default behaviour unchanged. However, if the general consensus is that keeping the window settings is the preferable default, I am very happy to change it

KoBeWi commented:
Instead of adding a new method, this can be done in `_save_docks()` (which contrary to its name, saves not only docks).

spanzeri commented:
Thank you @KoBeWi for the feedback. Your suggested change has the additional side-effect of saving the window position when docks are changed, which I see as a bonus (if the editor crashes or is otherwise killed, it would still remember the position). I have updated the code

trollodel commented:
@spanzeri I did something similar in #62378 for docks, here: https://github.com/godotengine/godot/pull/62378/files#diff-b88f0bde58b731e8dd353707c8011fb7a1e452f0db2c9f15f4b7905d60e5e52eR5230-R5265
You might want to take a look at it, especially the fallbacks.

spanzeri commented:
@trollodel first of all, I love your PR. I would love to have the script editor detachable. It is probably a good idea to add some checks on reload, as soon as I have some time I will improve the restore code to be a bit smarter

spanzeri commented:
@trollodel I had a look and I am conflicted about making adjustments.
The settings are read before both the display server and the editor node are created.
The editor_node is aware of the minimum size for a working editor and fixes the windows size during construction.
The position and screen are ensured by the display_driver implementations that all have some platform specific logic to deal with off-screen windows (or in some cases, like android, they ignore the settings altogether).
Your changelist was different on this aspect because it deals with windows created after the first one, which at the very least need a valid screen index and a valid size.

For this reason I think it would be better to not modify the initial window after the display server is created to not cause unexpected behaviours if either the editor_node or display server implementation is changed.
If you can think of a specific case where this breaks that I did not think of, I am happy to discuss it :)

trollodel commented:
Didn't know about the checks done in display servers.
In any case, I wouldn't rely on display server and its fallback behavior. For example, I would center the window in the screen if the saved rect goes off-screen.

However, I agree that the context in #62378 is different, so feel free to take whatever you think is useful and leave the rest.

spanzeri commented:
Implemented logic to re-fit the editor window to the screen usable rect or their position has changed since the last editor cfg was saved

bruvzg commented on main/main.cpp:
```diff
@@ -2021,6 +2052,30 @@ Error Main::setup2(Thread::ID p_main_tid_override) {
 		}
 	}
 
+#ifdef TOOLS_ENABLED
+	// If the editor is running in windowed mode, ensure the window rect fits
+	// the screen in case screen count or position has changed
+	if (editor && init_windowed) {
+		// We still need to check we are actually in windowed mode, because
+		// certain platform might only support one fullscreen window.
+		if (DisplayServer::get_singleton()->window_get_mode() == DisplayServer::WINDOW_MODE_WINDOWED) {
+			Vector2i current_size = DisplayServer::get_singleton()->window_get_size();
+			Vector2i current_pos = DisplayServer::get_singleton()->window_get_position();
+			int screen = DisplayServer::get_singleton()->window_get_current_screen();
+			Rect2i screen_rect = DisplayServer::get_singleton()->screen_get_usable_rect(screen);
+
+			Vector2i adjusted_end = screen_rect.get_end().min(current_pos + current_size);
+			Vector2i adjusted_pos = screen_rect.get_position().max(adjusted_end - current_size);
+			Vector2i adjusted_size = DisplayServer::get_singleton()->window_get_min_size().max(adjusted_end - adjusted_pos);
+
+			if (current_pos != adjusted_end || current_size != adjusted_size) {
+				DisplayServer::get_singleton()->window_set_position(adjusted_pos);
+				DisplayServer::get_singleton()->window_set_size(adjusted_size);
+			}
+		}
+	}
+#endif
```
It might be better to do it in the `DisplayServer` constructors instead, since doing it after can case already shown window to jump, or its size can be unnecessarily adjusted by OS/WM.

Relevant code:

https://github.com/godotengine/godot/blob/27253f3eb2c78a9ad5114c92eae2036b10e1d7e0/platform/windows/display_server_windows.cpp#L4209-L4217
https://github.com/godotengine/godot/blob/27253f3eb2c78a9ad5114c92eae2036b10e1d7e0/platform/linuxbsd/x11/display_server_x11.cpp#L5641-L5649
https://github.com/godotengine/godot/blob/27253f3eb2c78a9ad5114c92eae2036b10e1d7e0/platform/macos/display_server_macos.mm#L3878-L3886

spanzeri commented on main/main.cpp:
```diff
@@ -1977,6 +1977,37 @@ Error Main::setup2(Thread::ID p_main_tid_override) {
+
+	if (editor) {
+		bool restore_editor_window = GLOBAL_DEF("editor/run/remember_window_size_and_position", true);
+		if (restore_editor_window) {
```
This is a very good point. In fact, I should also check init_windowed on top of the 2 bools you mentioned for the "-w" command line.

spanzeri commented:
I'll see if I can make it happen

nbooba commented:
Do it please, it's so annoying that it maximizes the window every time, like leave it alone

Calinou commented:
@nbooba Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

spanzeri commented:
I have finally spent some time looking at this PR again. Apologies for the long break, but work left me with a lot less spare time as of late.
I have changed the saving code to not override the size when the window is something other than windowed.
Now, this is not a perfect solution.
If one closes the window maximized, the next run of the editor, the window will remember the last non maximized size.
However, if the size is changed, the window maximized, then the editor gets closed, the last non-maximised size is not saved.
I have tried adding the WM_RESIZE notification, but that seems to trigger on the window, but not on the editor node.
I would prefer to not further scatter this code on a whole bunch of different places.
If anyone with a better knowledge of the engine/editor than me has an idea how to further improve this code (i.e. capture window events and store the last non-maximised size every time it changes) I am open to suggestions :)

oxeron commented:
Any improvement on this feature ?

spanzeri commented:
Discussion around this PR has stagnated a bit and the code is fairly out of date at this point. I'd need to find the time to merge with latest with master and see if I can improve it enough to get it over the line.
if there is interest, I could certainly try to find the time to do that

Marenz commented:
> if there is interest

Let me just say, every time my project (window) loads, I think of you ;)

oxeron commented:
A lot of godot users will thank you, including me :)

AThousandShips commented:
Please don't bump without contributing significant new information. Use the :+1: reaction button on the first post instead. Feedback is appreciated but this doesn't really add anything and it just pings a lot of people needlessly

AThousandShips commented on doc/classes/EditorSettings.xml:
```diff
@@ -688,6 +688,9 @@
 			The preferred monitor to display the project manager.
 		</member>
+		<member name="interface/editor/remember_window_size_and_position" type="bool" setter="" getter="">
+			If [code]true[/code], the editor window will remember its size, position and which screen it was display on across restarts.
```
```suggestion
			If [code]true[/code], the editor window will remember its size, position, and which screen it was display on across restarts.
```
Please use [Oxford comma](https://docs.godotengine.org/en/latest/contributing/documentation/docs_writing_guidelines.html#use-the-oxford-comma-to-enumerate-anything).

AThousandShips commented on main/main.cpp:
```diff
@@ -2642,8 +2649,39 @@ Error Main::setup2() {
 		}
 
+		bool has_command_line_window_override = init_use_custom_pos || init_use_custom_screen || init_windowed;
+		if (editor && !has_command_line_window_override) {
+			if (restore_editor_window_layout) {
+				Ref<ConfigFile> config;
+				config.instantiate();
+				// Load and amend existing config if it exists.
+				Error err = config->load(EditorPaths::get_singleton()->get_project_settings_dir().path_join("editor_layout.cfg"));
+				if (err == OK) {
+					init_screen = config->get_value("EditorWindow", "screen", init_screen);
+					String mode = config->get_value("EditorWindow", "mode", "maximized");
+					window_size = config->get_value("EditorWindow", "size", window_size);
+					if (mode == "windowed") {
+						window_mode = DisplayServer::WINDOW_MODE_WINDOWED;
+						init_windowed = true;
+					} else if (mode == "fullscreen") {
+						window_mode = DisplayServer::WINDOW_MODE_FULLSCREEN;
+						init_fullscreen = true;
+					} else {
+						window_mode = DisplayServer::WINDOW_MODE_MAXIMIZED;
+						init_maximized = true;
+					}
+
+					if (init_windowed) {
+						init_use_custom_pos = true;
+						init_custom_pos = config->get_value("EditorWindow", "position", Vector2i(0, 0));
+					}
+				}
+			}
```
```suggestion
		if (editor && !has_command_line_window_override && restore_editor_window_layout) {
			Ref<ConfigFile> config;
			config.instantiate();
			// Load and amend existing config if it exists.
			Error err = config->load(EditorPaths::get_singleton()->get_project_settings_dir().path_join("editor_layout.cfg"));
			if (err == OK) {
				init_screen = config->get_value("EditorWindow", "screen", init_screen);
				String mode = config->get_value("EditorWindow", "mode", "maximized");
				window_size = config->get_value("EditorWindow", "size", window_size);
				if (mode == "windowed") {
					window_mode = DisplayServer::WINDOW_MODE_WINDOWED;
					init_windowed = true;
				} else if (mode == "fullscreen") {
					window_mode = DisplayServer::WINDOW_MODE_FULLSCREEN;
					init_fullscreen = true;
				} else {
					window_mode = DisplayServer::WINDOW_MODE_MAXIMIZED;
					init_maximized = true;
				}

				if (init_windowed) {
					init_use_custom_pos = true;
					init_custom_pos = config->get_value("EditorWindow", "position", Vector2i(0, 0));
				}
			}
```

AThousandShips commented on editor/editor_node.h:
```diff
@@ -490,6 +490,9 @@ class EditorNode : public Node {
 	bool run_surface_upgrade_tool = false;
 
+	bool has_valid_last_windowed_size = false;
+	Size2i last_windowed_size = Size2i(0, 0);
```
```suggestion
	Size2i last_windowed_size;
```
Unnecessary default

AThousandShips commented on main/main.cpp:
```diff
@@ -2762,6 +2800,30 @@ Error Main::setup2() {
 
+#ifdef TOOLS_ENABLED
+	// If the editor is running in windowed mode, ensure the window rect fits
+	// the screen in case screen count or position has changed
```
```suggestion
	// the screen in case screen count or position has changed.
```

KoBeWi replied:
```suggestion
	// The screen in case screen count or position has changed.
```

spanzeri commented:
New and improved (hopefully) version of this PR.
I made a few changes to address some of the standing issues/concerns.

- The editor node now detects window size changes and can remember the last state (maximized, full-screen or windowed) even if the editor is closed while minimised;
- I moved the code within the trace events so it is benchmarked with the other startup code;
- I moved the setting so it is now a editor config rather than a project config (note that position, size, mode and screen are still saved per project so people can have different editor config for different projects)

This is on by default in the current version, but if we'd rather have it off  I am happy to flip the condition.
I have also retested the ability of the window to recenter/resize if it starts off-screen (for example in case of a monitor being unplugged) and that seems to be working fine.

KoBeWi commented on main/main.cpp:
```diff
@@ -2642,8 +2649,37 @@ Error Main::setup2() {
+
 		OS::get_singleton()->benchmark_end_measure("Startup", "Initialize Early Settings");
 	}
+
```
```suggestion
```
Unnecessary.

KoBeWi commented on editor/editor_settings.cpp:
```diff
@@ -450,6 +450,7 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
 	_initial_set("interface/editor/separate_distraction_mode", false);
 	_initial_set("interface/editor/automatically_open_screenshots", true);
 	EDITOR_SETTING_USAGE(Variant::BOOL, PROPERTY_HINT_NONE, "interface/editor/single_window_mode", false, "", PROPERTY_USAGE_DEFAULT | PROPERTY_USAGE_RESTART_IF_CHANGED)
+	EDITOR_SETTING(Variant::BOOL, PROPERTY_HINT_NONE, "interface/editor/remember_window_size_and_position", true, "")
```
```suggestion
	_initial_set("interface/editor/remember_window_size_and_position", true)
```

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -5187,6 +5205,38 @@ void EditorNode::_load_central_editor_layout_from_config(Ref<ConfigFile> p_confi
+				p_layout->set_value(p_section, "mode", "fullscreen");
+				break;
+			case Window::MODE_MINIMIZED:
+				if (has_valid_last_windowed_size) {
```
When this can be true?

KoBeWi commented on editor/editor_node.cpp:
```diff
@@ -1181,6 +1184,20 @@ void EditorNode::_reload_project_settings() {
+	if (w) {
+		Window::Mode mode = w->get_mode();
+		if (mode == Window::MODE_WINDOWED) {
+			has_valid_last_windowed_size = true;
```
This value is unused.

KoBeWi commented:
This works correctly now, but there are some weird things in the implementation. `has_valid_last_windowed_size` has effectively no effect. Also is there a reason why you connect to `size_changed` signal instead of just fetching values when saving layout? This seems to have to do something with minimized window, but it's implemented incorrectly and does nothing. If editor is closed with minimized window, IMO it's fine if it's stored as fullscreen.

Or you can save the window as minimized, so it opens minimized 😏

spanzeri commented:
Thank you @KoBeWi  for the review.
I have removed the empty line and removed the last window size as you suggested, because was indeed not needed.
However, the signal and bool were working as intended.
They are there- as you guessed - to deal with the case when the editor window is closed while minimised.
With the current implementation (which is hopefully a bit clearer), we keep track if the window was windowed before being minimised. On the next startup, we can restore the window to the last size and windowed mode.

I could save maximised in that case, but imho remembering the last user config is a more pleasant behaviour than maximising.

I could also store the window mode as minimised and start the editor minimised on restart, but that would just be evil :)

I have run a few tests again and the bool is changing as expected, but if there is something you see and I am missing feel free to let me know.

As a side node, I'd be preferable to not have to connect the signal and rely on notification, but unfortunately WM events seems to only fire for windows and viewports and when I added a printf to _notification there was nothing coming through for minimise/restore event on the editor node.
But if there is a better way to handle those cases that I don't know about, I'd be happy to make another change

AThousandShips commented on doc/classes/EditorSettings.xml:
```diff
@@ -688,6 +688,9 @@
 			The preferred monitor to display the project manager.
 		</member>
+		<member name="interface/editor/remember_window_size_and_position" type="bool" setter="" getter="">
+			If [code]true[/code], the editor window will remember its size, position, and which screen it was display on across restarts.
```
```suggestion
			If [code]true[/code], the editor window will remember its size, position, and which screen it was displayed on across restarts.
```

Calinou reviewed (commented):
Tested locally on Linux/X11 (rebased on top of `master` https://github.com/godotengine/godot/commit/1567a498cbe0d2aeb73a9d2a67d3ac6b22b340a1), I noticed some issues:

- Window position/size is only remembered if you press <kbd>Ctrl + S</kbd> before exiting the editor. I think it should always be remembered, even if you exit the editor without saving changes or even if the editor crashes (or is exited with <kbd>Ctrl + C</kbd> from a terminal).
- Window position is restored correctly, but size is not. Only the first saved size is taken into account, and is reused afterwards.

On the plus side, this PR seems to completely fix the startup focus/size change issues that were present for years. The window position/size is only set once and the window doesn't forcibly move to the foreground once the editor is done loading, which is a big usability upgrade when opening lots of projects.

KoBeWi commented:
> Window position/size is only remembered if you press Ctrl + S before exiting the editor.
> Window position is restored correctly, but size is not.

That wasn't the case when I tested 🤔

spanzeri commented:
@Calinou that is very surprising indeed. The function that saves position and size is called by the same function that stores layout, and always on close (without the need for saving). Also, the size and position are stored at the same time.
Is it possible you had some command line overrides?  My CL intentionally does nothing if you specify a command line argument as I assume you know what you want and you don't want me to mess with it.
If you don't, then I need to figure out why you are seeing that issue (I'm also on linux nowadays, but wayland and not X11)

spanzeri commented:
I have rebased with latest master and run the change again.
On linux + wayland I can confirm it saves both position as size as expected (without the need for saving).
Regarding save on Ctrl+C, that is a bit more complicated. I am not sure if we have some signal handler, but it would need to happen on a OS specific manner and I am not even sure the window would still be available at that point.
However, I do feel like there is less of an expectation for the window size and position to be remembered on those circumstances (crash, sigabort/sigkill, terminate from a debug session etc).
I know Unreal for example also does not save window position and size on those cases

Calinou commented:
> Is it possible you had some command line overrides?

No, I started the editor from the command line (by specifying the path to the `project.godot` file) but that's about it.

spanzeri commented:
Would you be able to open editor_layout.cfg from the cache directory of your project (path_to_project/.godot/editor/editor_layout.cfg) and see what's inside there? Should look something like this
![image](https://github.com/godotengine/godot/assets/494097/2e60d6ae-738b-444c-8798-a1297179f5a7)
I also just realized the save function is called on resize (due to the layout being saved on resize). Maybe it should also call (at least the new function) on move, so even if the editor crashes it will have saved both correctly if you moved/resized

Calinou reviewed (approved):
Working great now :slightly_smiling_face: 

I've also tested scenarios where I change the monitor resolution while the editor is closed and it works as expected (maximized mode is retained).

The only issue I noticed is that when you start the editor maximized, if you un-maximize it or drag its window during the startup process, it will maximize itself again once the editor is done starting. This does not occur the other way around (starting non-maximized and maximizing during the startup process).

Either way, this got to a good enough state where I consider this mergeable.

KoB-Kirito commented:
Will this get merged soon? My Godot always opens in fullscreen no matter how I close it and it's so annoying :)

akien-mga commented:
Thanks for the great usability improvement, and for sticking with the long review process!

som45oul commented:
Really glad to see this change has been merged, and looking forward to it making its way into release.

Is there any possibility that the size and position could also be added as command line options, perhaps?  Assuming you have cracked the nut of being able to restore the window based on a configuration / project file, perhaps there could also be a command-line override?

Reason for this is wanting to "force" particular sizes and positions for new projects.  Unless your change already accounts for that, and I just misunderstood?

spanzeri commented:
@som45oul command line option for size and other display properties have been present for a while. See here: [https://docs.godotengine.org/en/stable/tutorials/editor/command_line_tutorial.html#command-line-reference](Command Line Reference)

som45oul commented:
Thanks @spanzeri. I believe I read somewhere these options had been deprecated and no longer work. I tested this previously and found that to be the case, but I will check again. Appreciate the reply.

LouChiSoft commented:
Hey, was this merge reversed at somepoint? I see the changes were merged directly into the `master` branch. But building the master branch doesn't have an option in the project settings and insepcting EditorSettings.xml also shows no such setting

KoBeWi commented:
The setting is now part of `interface/editor/editor_screen` as "Auto (Remembers Last Position)".

LouChiSoft commented:
Excellent, found it. Thanks

danijmn commented on main/main.cpp:
```diff
@@ -2768,6 +2803,30 @@ Error Main::setup2(bool p_show_boot_logo) {
+			Vector2i adjusted_pos = screen_rect.get_position().max(adjusted_end - current_size);
+			Vector2i adjusted_size = DisplayServer::get_singleton()->window_get_min_size().max(adjusted_end - adjusted_pos);
+
+			if (current_pos != adjusted_end || current_size != adjusted_size) {
```
Did you mean `if (current_pos != adjusted_pos ...` instead of `adjusted_end`?
